calyx-py/test/correctness/arbiter_6.py

```diff
@@ -106,59 +30,64 @@ def add_wrap2(prog):
     j_mod_4 = wrap.reg("j_mod_4", 32)
 
     # Additional cells and groups to compute equality and lt
-    eq0cell, eq0grp = add_eq(wrap, i, 0, "eq0", "i_eq_0")
-    eq1cell, eq1grp = add_eq(wrap, i, 1, "eq1", "i_eq_1")
-    lt1cell, lt1grp = add_lt(wrap, j, 4, "lt1", "j_lt_4")
-    lt2cell, lt2grp = add_lt(wrap, j, 8, "lt2", "j_lt_8")
+    i_eq_0_cell, i_eq_0_grp = util.insert_eq(wrap, i, 0, "i_eq_0", 32)
+    i_eq_1_cell, i_eq_1_group = util.insert_eq(wrap, i, 1, "i_eq_1", 32)
+    j_lt_4_cell, j_lt_4_group = util.insert_lt(wrap, j, 4, "j_lt_4", 32)
+    j_lt_8_cell, j_lt_8_group = util.insert_lt(wrap, j, 8, "j_lt_8", 32)
 
     # Load `j` unchanged into `j_mod_4`.
-    unchanged = add_reg_load(wrap, j, j_mod_4, "j_unchanged")
+    unchanged = util.insert_reg_store(wrap, j_mod_4, j, "j_unchanged")
 
-    # A subtraction cell and wiring to perform j-4 and j-8.
-    sub_cell = wrap.sub("sub", 32)
-    sub1cell = add_sub(wrap, j, cb.const(32, 4), sub_cell, j_mod_4, "j_minus_4")
-    sub2cell = add_sub(wrap, j, cb.const(32, 8), sub_cell, j_mod_4, "j_minus_8")
+    # Wiring to perform j-4 and j-8. Either of these will store the result in `j_mod_4`.
+    j_minus_4 = util.insert_sub_and_store(
+        wrap, j, cb.const(32, 4), "j_minus_4", 32, j_mod_4
+    )
+    j_minus_8 = util.insert_sub_and_store(
+        wrap, j, cb.const(32, 8), "j_minus_8", 32, j_mod_4
+    )
 
     load_from_mems = [
         # Add wiring to load the value `j_mod_4` from all of the memory cells.
         # We'll have to invoke the correct one of these groups later on.
-        add_mem_load(wrap, mems[i], j_mod_4.out, ans, f"load_from_mem{i}")
+        util.insert_mem_load_to_mem(
+            wrap, mems[i], j_mod_4.out, ans, cb.const(32, 0), f"load_from_mem{i}"
+        )
         for i in range(6)
     ]
 
     wrap.control += [
         cb.if_(
-            lt1cell.out,
-            lt1grp,
+            j_lt_4_cell.out,
```

Comment:
Then, to an `if` or a `while`, I immediately pass
- `cell.out`
- `group`
- some `then_body`
- Optionally, some `else_body`

Could we not just make a slightly sugary variant of `if` (resp. `while`) that takes a cell-group tuple `cellgroup` as its first argument, and just calls the "normal" `if` (resp. `while`) with `cellgroup[0].out` and `cellgroup[1]` as the first two arguments?

Reply:
Yeah, I think this isn't a terrible idea. We should make some sort of struct that represents a port that is only meaningful when the provided combinational group is active.

Reply:
Indeed; I came here to say the same thing: this is a good idea, and we could introduce a small Python class that takes the place of the tuple. It could have a couple of convenience methods to get, for example, the `out` port of the relevant combinational component. Then the alternate forms of `if_`/`while_` would accept that object instead of separate port/group values. (Maybe for a future PR!)

Reply:
Moved to its own issue: #1638! Resolving this conversation so we can move towards merging this PR